File: backend/app/schemas/analytics.py
"""
IntelliHall — Hall Analytics Pydantic Schemas

Response shapes for GET /halls/{hall_id}/analytics.
"""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field

from app.models.enums import ComplaintCategory, ComplaintStatus
from app.schemas.complaint import ComplaintSummary


class AnalyticsSummaryRead(BaseModel):
    """High-level KPI counts for a hall."""

    total_complaints: int = Field(..., description="All complaints ever filed in this hall.")
    open_complaints: int = Field(
        ...,
        description="Complaints not yet closed (status != closed).",
    )
    in_progress: int = Field(..., description="Complaints with status in_progress.")
    completed: int = Field(..., description="Complaints with status closed.")
    critical_complaints: int = Field(
        ...,
        description="Active critical-priority complaints (priority=critical, not closed).",
    )
    complaints_today: int = Field(
        ...,
        description="Complaints created since midnight UTC today.",
    )
    complaints_this_week: int = Field(
        ...,
        description="Complaints created since Monday 00:00 UTC this week.",
    )
    avg_resolution_time_hours: float | None = Field(
        default=None,
        description=(
            "Average hours from complaint creation to first closed status transition. "
            "Null when no closed complaints exist."
        ),
    )


class CategoryCountRead(BaseModel):
    """Complaint count grouped by maintenance category."""

    category: ComplaintCategory
    count: int


class StatusCountRead(BaseModel):
    """Complaint count grouped by lifecycle status."""

    status: ComplaintStatus
    count: int


class MonthlyTrendPointRead(BaseModel):
    """Complaint volume for a single calendar month."""

    month: str = Field(..., description="Month key in YYYY-MM format.")
    count: int


class HallAnalyticsRead(BaseModel):
    """Full analytics payload for a hall administrator dashboard."""

    hall_id: str
    hall_name: str
    summary: AnalyticsSummaryRead
    by_category: list[CategoryCountRead]
    by_status: list[StatusCountRead]
    monthly_trend: list[MonthlyTrendPointRead]
    recent_complaints: list[ComplaintSummary]

File: frontend/src/types/analytics.ts
import { ComplaintCategory, ComplaintStatus, ComplaintSummary } from "./complaint";

export interface AnalyticsSummary {
  total_complaints: number;
  open_complaints: number;
  in_progress: number;
  completed: number;
  critical_complaints: number;
  complaints_today: number;
  complaints_this_week: number;
  avg_resolution_time_hours: number | null;
}

export interface CategoryCount {
  category: ComplaintCategory;
  count: number;
}

export interface StatusCount {
  status: ComplaintStatus;
  count: number;
}

export interface MonthlyTrendPoint {
  month: string;
  count: number;
}

export interface HallAnalytics {
  hall_id: string;
  hall_name: string;
  summary: AnalyticsSummary;
  by_category: CategoryCount[];
  by_status: StatusCount[];
  monthly_trend: MonthlyTrendPoint[];
  recent_complaints: ComplaintSummary[];
}
